File: java/src/com/alofmethbin/jadawin/Compiler.java
package com.alofmethbin.jadawin;

import com.alofmethbin.jadawin.elements.Element;
import com.alofmethbin.jadawin.elements.Error;
import com.alofmethbin.jadawin.elements.Expect;
import com.alofmethbin.jadawin.elements.Image;
import com.alofmethbin.jadawin.elements.Tag;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.BufferedReader;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import gg.jte.CodeResolver;
import gg.jte.ContentType;
import gg.jte.TemplateEngine;
import gg.jte.TemplateOutput;
import gg.jte.output.StringOutput;
import gg.jte.resolve.DirectoryCodeResolver;
import java.io.File;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.lang.reflect.Constructor;
import java.nio.file.Path;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class Compiler {
    private List<String> errors = new ArrayList<>();
    private File source, sink;
    private Any config;
    
    private Set<String> words, names, newWords;
    private Set<String> generated = new HashSet<>();
    private Map<String, List<Page>> key2Pages = new HashMap<>();
    private TemplateEngine templateEngine;
    private byte[] fromBuffer = new byte[1000000];
    private byte[] toBuffer = new byte[1000000];
    private LocalDate now   = LocalDate.now();
    private static Pattern assetFilename = Pattern.compile( "\\.(JPG|JPEG|jpg|jpeg|png|zip|gif|svg|webp)$");
    private static Pattern directiveLine = Pattern.compile( "^@(\\S*)(.*)$");
    private static Pattern htmlEntities  = Pattern.compile( "&(.)(acute|caron|cedil|grave|circ|slash|uml|ring);");
    private static Pattern quoteStart    = Pattern.compile( "^(['\"]*)(.*)$");
    private static Pattern quoteEnd      = Pattern.compile( "^(.*)(['\"]*)$");
    private static Pattern prefixStart   = Pattern.compile( "^(pre|quasi|ex|half|mini|multi|non)-(.*)$");

    public Compiler(String source, String configPath, String sink) throws Exception {
        this.source = new File(source);
        this.sink = new File(sink);
        ObjectMapper mapper = new ObjectMapper( new YAMLFactory());
        config = new Any( mapper.readValue( new File( configPath), Map.class));
        loadWords();
        CodeResolver codeResolver = new DirectoryCodeResolver(Path.of( config.get( "jte").asString())); 
        templateEngine = TemplateEngine.create(codeResolver, ContentType.Plain); 
    }

    private void addContent(String verb, Article article, List<String> info, String ref) {
        Class clazz;
        try {
            clazz = Class.forName("com.alofmethbin.jadawin.elements." + Utils.capitalise( verb));
        } catch (ClassNotFoundException cnfe) {
            article.error("Unknown directive: " + verb);
            return;
        }

        Constructor construct;
        try {
            construct = clazz.getConstructor(new Class[]{Article.class, List.class});
        } catch (NoSuchMethodException nsme) {
            article.error("Bad directive: " + verb);
            return;
        }

        Element element;
        try {
            element = (Element) construct.newInstance(new Object[]{article, info});
        } catch (Exception ex) {
            String msg = ex.getMessage();
            article.error( (msg != null) ? msg : ex.getClass().getName());
            return;
        }

        if (! element.hasError()) {
            article.addContent(element, info.size() > 1);
        }
    }
    
    public void compile() throws Exception {
        out("... Initialised");
        Image.findImages(this.source);
        Article homePage = parseDirectory(null, "");
        out("... Parsed");
        homePage.prepare();
        out("... Prepared");
        regenerate(homePage);
        out("... Generated");
        //tidyUp(this.sink);
        out("... Tidied up");
        Tag.checkAllRendered();
        out("... Check tags all used");
        reportNewWords();
        out("... " + (Expect.numberOfTests() + Error.numberOfTests()) + " tests run");
        Error.reportErrors();
        for (String msg : errors) {
            out("*** " + msg);
        }
        if ( hasErrors() ) {
            out("*** " + errors.size() + " errors");
        }
    }

    public void copyResource(String path) throws Exception {
        File from = new File(source, path);
        File to = new File(sink, path);

        if (!to.getParentFile().exists()) {
            to.getParentFile().mkdirs();
        }

        record(to);

        int fromLen = loadBinary(from, fromBuffer);
        if (to.exists()) {
            int toLen = loadBinary(to, toBuffer);
            boolean same = (fromLen == toLen);
            for (int i = 0; same && (i < fromLen); i++) {
                same = (fromBuffer[i] == toBuffer[i]);
            }

            if (same) {
                return;
            }
        }

        try ( FileOutputStream os = new FileOutputStream(to)) {
            os.write(fromBuffer, 0, fromLen);
        } catch (Exception ex) {
            out("*** Error copying resource to " + to);
            throw ex;
        }
    }

    public int [][] dimensions( String key) {
        return config.get( "dimensions").get(key).asInts2();
    }
    
    public synchronized void error( String path, String msg) {
        errors.add( msg + " [" + path + "]");
    }

    public String getRootURL() {
        return config.get("root_url").asString();
    }

    private boolean hasErrors() {
        return ! errors.isEmpty();
    }
    
    public boolean isOffsite( String url) {
        if ( url.startsWith( getRootURL()) ) {
            return false;
        } else if ( Utils.isAbsoluteUrl( url) ) {
            return true;
        } else {
            return false;
        }
    }
    
    private int loadBinary(File from, byte[] fromBuffer) throws Exception {
        throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
    }

    private void loadWords() {
        String spellChecking = config.get("spell_checking").asString();
        if (spellChecking  == null) {return;}
        
        File wordsDir = new File( spellChecking);
        if (! (wordsDir.exists() && wordsDir.isDirectory())) {
            errors.add( "Not a directory: spell_checking setting in config");
            return;
        }
        
        this.words    = new HashSet<>();
        this.names    = new HashSet<>();
        this.newWords = new HashSet<>();
        
        for (File f: wordsDir.listFiles()) {
            if ( f.getName().endsWith( ".txt") ) {
                try (BufferedReader reader = new BufferedReader( new FileReader( f))) {
                    String line;
                    while ((line = reader.readLine()) != null) {
                        String word = line.trim();
                        if ( word.isEmpty() ) {continue;}
                        
                        if ( word.equals( word.toLowerCase()) ) {
                            words.add( word.toLowerCase());
                        } else {
                            names.add( word);
                        }
                    }
                } catch (Exception ex) {
                    errors.add( f.getAbsolutePath() + ": " + ex.getMessage());
                }
            }
        }
    }

    public Page lookupPage( Page referrer, String path, String ref) {
        List<Page> matches = this.key2Pages.get( path);
        if (matches == null) {
            referrer.error( "Path not found for " + ref);
            return null;
        } else if (matches.size() > 1) {
            referrer.error( "Path not found for " + ref);
            return null;
        } else {
            return matches.get(0);
        }
    }
    
    public int [] menuItemsPerLine() {
        return config.get("menu_items_per_line").asInts();
    }

    public LocalDate now() {
        return now;
    }

    public int numMenuStyles() {
        return config.get("num_menu_styles").asInt();
    }
    
    private void out(String msg) {
        System.out.println(msg);
    }

    private Article parseDirectory( Article parent, String dir) {
        File dirFile = dir.equals("") ? sink : new File( sink, dir);
        if (! dirFile.exists()) {dirFile.mkdirs();}
        
        Article dirArticle = new Article( this, parent, dir + "/index.txt");
        remember( dir.isEmpty() ? "." : dir, dirArticle);
        
        if (parent != null) {
            parent.addChild( dirArticle);
        }
        
        for (File f: (new File( source, dir)).listFiles()) {
            if ( f.getName().startsWith( ".") ) {continue;}
            if ( f.getName().startsWith( "_") ) {continue;}
            String path = dir + "/" + f.getName();
            
            if ( f.isDirectory() ) {
                parseDirectory( dirArticle, path);
            } else if ( f.getName().equals( "index.txt") ) {
                parseFile( dirArticle, f);
            } else if ( f.getName().endsWith( ".txt") ) {
                Article a = new Article( this, dirArticle, path);
                dirArticle.addChild( a);
                remember( path, a);
                parseFile( a, f);
            } else if ( assetFilename.matcher( f.getName()).find() ) {
                //remember( path, new File( sink, path));
            } else {
                dirArticle.error( "Unhandled file: " + f.getName());
            }
        }
        
        dirArticle.discardFutureChildren();
        return dirArticle;
    }

    private void parseFile( Article article, File f) {
        try (BufferedReader reader = new BufferedReader( new FileReader( f))) {
            List<String> lines = new ArrayList<>();
            String line, line1;
            while ((line = reader.readLine()) != null) {
               Matcher m = directiveLine.matcher( line);
               if ( m.matches() ) {
                   lines.clear();
                   if (! m.group(2).isBlank()) {
                       lines.add( m.group(2).trim());
                       addContent( m.group(1), article, lines, line);
                   } else {
                       reader.mark( 1000);
                       int lastNonWhite = 0;
                       while ((line1 = reader.readLine()) != null) {
                           if ( directiveLine.matcher( line1).find() ) {
                               reader.reset();
                               break;
                           } else {
                               reader.mark( 1000);
                               lines.add( line1);
                               if (! line1.isBlank()) {
                                   lastNonWhite = lines.size();
                               }
                           }
                       }
                       addContent( m.group(1), 
                                   article, 
                                   lines.subList( 0, lastNonWhite), 
                                   line);
                   }
               } else if (! line.isBlank()) {
                   article.error( "Expected directive: " + line);
                   break;
               }
            }
        } catch (Exception ex) {
            article.error( ex.getMessage());
        }
    }

    public void record( File to) {
        this.generated.add( to.getAbsolutePath());
    }

    private void regenerate( Article article) throws Exception {
        regenerate( config.get( "colours").asStringMap(),
                    config.get( "dimensions").asInts2Map(),
                    article);
    }
    
    private void regenerate( Map<String,String> colours,
                             Map<String,int [][]> dimensions,
                             Article article) throws Exception {
        TemplateOutput output = new StringOutput();
        Map<String,Object> args= new HashMap<>();
        args.put( "colours", colours);
        args.put( "dimensions", dimensions);
        args.put( "article", article);
        this.templateEngine.render( "page.jte", args, output);
        String html = output.toString();
        html = article.postProcessHTML(html);
        
        File sinkFile = article.getSinkFile();
        generated.add( sinkFile.getCanonicalPath());
        
        if ( sinkFile.exists() ) {
            String old = Utils.read( sinkFile);
            if (! old.trim().equals(html.trim())) {
                Utils.write(html, sinkFile);
            }
        } else {
            Utils.write( output.toString(), sinkFile);
        }
        
        for (var page: article.children()) {
            if (page instanceof Article article1) {
                regenerate( colours, dimensions, article1);
            }
        }
    }

    private void remember( String path, Page page) {
        remember1( path, page);
        for (int i = 0; i < path.length() - 1; i++) {
            if ( path.substring(i, i+1).equals( "/") ) {
                remember1( path.substring(i+1), page);
            }
        }
    }

    private void remember1( String path, Page page) {
        path = path.replaceFirst( "\\.txt$", "");
                
        if (! key2Pages.containsKey( path)) {
            List<Page> list = new ArrayList<>();
            key2Pages.put( path, list);
            list.add( page);
        } else {
            key2Pages.get( path).add( page);
        }
    }

    private void reportNewWords() {
        String misspeltWords = config.get( "misspelt_words").asString();
        if (misspeltWords != null) {
            try (FileWriter w = new FileWriter( misspeltWords)) {
               for (String word: newWords) {
                   w.write( word + "\n");
               }
            } catch (Exception ex) {
               System.err.println( "Error writing unknown words: " + ex.getMessage());
            }
        }
    }

    public File sourceFile( String path) {
        return new File( source, path);
    }

    public void spellCheck( Article article, String text) {
        if (this.words == null) {return;}
        
        for (String part: text.split( "[,\\.!\\? \\n\\(\\);:]")) {
            String word = part;
            
            Matcher m = Compiler.quoteStart.matcher( word);
            if ( m.find() ) {
                word = m.group(2);
            }
            
            m = Compiler.prefixStart.matcher( word);
            if ( m.find() ) {
                word = m.group(2);
            }
            
            m = Compiler.quoteEnd.matcher( word);
            if ( m.find() ) {
                word = m.group(1);
            }
            
            m = Compiler.htmlEntities.matcher( word);
            while ( m.find() ) {
                word = word.substring( 0, m.start()) + 
                       m.group(1) + 
                       word.substring( m.end(), word.length());
                m    = Compiler.htmlEntities.matcher( word);
            }
            
            if ( this.words.contains( word.toLowerCase()) ) {continue;}
            if ( this.names.contains( word) ) {continue;}
            
            this.newWords.add( word);
            article.error( "Unknown word: " + part);
        }
    }
    
    private boolean tidyUp( File dir) throws IOException {
        boolean keep = false;
        for (File child: dir.listFiles()) {
            if ( child.isDirectory() ) {
                if ( tidyUp( child) ) {
                    keep = true;
                } else {
                    out( "Deleting " + toPath(child));
                    child.delete();
                }
            } else if ( child.getName().startsWith( ".") ) {
                child.delete();
            } else if ( this.generated.contains( child.getAbsolutePath()) ) {
                keep = true;
            } else {
                out( "Deleting " + toPath(child));
                child.delete();
            }
        }
        return keep;
    }

    public String toPath( File file) throws IOException {
        return Utils.toPath( sink, file);
    }

    public File toSinkFile( String path) {
        return new File( sink, path);
    }

    public File toSourceFile( String path) {
        return new File( source, path);
    }

    public static void main(String[] args) {
        try {
            Compiler c = new Compiler( args[0], args[1], args[2]);
            c.compile();
            if ( c.hasErrors() ) {
                System.exit(1);
            }
        } catch (Exception ex) {
            ex.printStackTrace();
            System.exit(1);
        }
    }
}

File: ruby/Compiler.rb
=begin
	Compiler.rb

	Generate HTML structure from article files
=end

require 'bundler/setup'
# require 'rexml/document'
# require 'rexml/xpath'
require 'yaml'
require 'cgi'
require 'fileutils'
require 'json'

require_relative 'Article'
require_relative 'elements/error'
require_relative 'utils'

class Compiler
  include Utils

  # Initialisation
  def initialize( source, config, sink, debug_pages=nil)
    @errors        = 0
    @source        = source
    @sink          = sink
    @debug_pages   = debug_pages.nil? ? nil : Regexp.new( debug_pages)
    @config        = YAML.load( File.open( config))
    @words, @names = load_words
    @new_words     = []
    @generated     = {}
    @key2paths     = Hash.new {|h,k| h[k] = []}
    prepare_templates
  end

# =================================================================
# Main logic
# =================================================================

  def add_content( verb, article, info, ref)
    begin
      require_relative( 'elements/' + verb.downcase)
    rescue LoadError
      article.error( 'Unknown directive: ' + verb)
      return
    end

    if verb.downcase == verb
      verb = verb.split( '_').collect {|n| n.capitalize}.join('')
    end

    begin
      clazz = Kernel.const_get( 'Elements::' + verb)
    rescue NameError
      article.error( 'Directive error: ' + ref)
      return
    end

    begin
      article.add_content( clazz.new( self, article, info), info.size > 1)
    rescue NoMethodError
      article.error( 'Directive error: ' + ref)
    end
  end

  # Compile
  def compile
    puts "... Initialised #{Time.now.strftime( '%Y-%m-%d %H:%M:%S')}"

    # Scan for images
    Elements::Image.find_images( @source)

    # Parse all the articles recursively
    @articles = parse( nil, "")
    puts "... Parsed      #{Time.now.strftime( '%Y-%m-%d %H:%M:%S')}"

    # Prepare the articles now all articles parsed
    prepare( [], @articles)
    puts "... Prepared    #{Time.now.strftime( '%Y-%m-%d %H:%M:%S')}"

    # Regenerate the HTML files
    regenerate( [], @articles)
    puts "... Generated   #{Time.now.strftime( '%Y-%m-%d %H:%M:%S')}"

    # Delete files not regenerated
    tidy_up( @sink)

    # Check locations all rendered
    Elements::Tag.check_all_rendered

    # Report new words
    report_new_words

    # Report tests run
    puts "... #{Elements::Expect.number_of_tests + Elements::Error.number_of_tests} tests run"

    # Report any errors
    Elements::Error.report_errors
    report_errors( @articles)
    puts "*** #{@errors} Errors in compilation" if @errors > 0
  end

  def copy_resource( f)
    from   = @source + '/' + f
    to     = @sink + '/' + f
    to_dir = File.dirname( to)
    Dir.mkdir( to_dir) unless File.exist?( to_dir)
    record( to)
    unless File.exist?( to) && (IO.read( to) == IO.read( from))
      FileUtils.cp( from, to)
    end
  end

  def debug_hook( article, msg)
    # puts "DEBUG100: #{article.sink_filename}"
    if @debug_pages && (@debug_pages =~ article.filename)
      puts "!!! #{article.filename} : #{msg}"
    end
  end

  # Parse the articles
  def parse( parent, path)

    # Article for the directory
    dir_article = Article.new( path + '/index.html')
    remember( (path == '') ? '.' : path, dir_article)
    parent.add_child( dir_article) if parent

    # Hash of articles in this directory
    dir_articles = Hash.new do |h,k|
      a = Article.new( path + '/' + k + '.html')
      dir_article.add_child( a)
      remember( path + '/' + k, a)
      h[k] = a
    end
    dir_articles['index'] = dir_article

    # Loop over source files - skip image files and other specials
    Dir.entries( @source + path).each do |file|
      next if /^\./ =~ file
      next if /^\_/ =~ file
      path1 = path + "/" + file

      if File.directory?( @source + path1)
        Dir.mkdir( @sink + path1) if not File.exist?( @sink + path1)
        parse( dir_article, path1)
      elsif m = /^(.*)\.txt$/.match( file)
        child = dir_articles[m[1]]
        parse_article(path, file, child)
      elsif /\.(JPG|jpg|JPEG|jpeg|png|zip|rb|kml|afphoto|command|erb|pdf|gif|svg|webp)$/ =~ file
        remember( path1, @source + path1)
      else
        raise "Unhandled file: #{path1}"
      end
    end

    dir_article.discard_future_children
  end

  def parse_article(path, file, article)
    debug_hook( article, "Parsing article")
    text = IO.read( @source + path + "/" + file)
    lines = text.split( "\n")
    i = 0

    while i < lines.size
      if m = /^@(\S+)\s*$/.match( lines[i])
        j = i
        while ((j+1) < lines.size) && (! (/^@\S/ =~ lines[j+1]))
          j = j + 1
        end

        k = j
        while (k > (i+1)) && (lines[k].strip == '')
          k = k - 1
        end

        add_content( m[1], article, lines[(i+1)..k], lines[i])
        i = j + 1
      elsif m2 = /^@(\S+)\s+(\S.*)$/.match( lines[i])
        add_content( m2[1], article, [m2[2].strip], lines[i])
        i += 1
      elsif lines[i].strip == ''
        i = lines.size + 1
      else
        article.error( "Expected directive: " + lines[i])
        i = lines.size + 1
      end
    end
  end

  def prepare_templates
    if @config['mode'].nil? ||
        @config['mode'] == 'liquid' ||
        @config['mode'] == 'liquid-c'
      require 'liquid'
      if @config['mode'] == 'liquid-c'
        require 'liquid/c'
      end
      Liquid::Template.file_system = Liquid::LocalFileSystem.new( @config['liquid'],
                                                                  pattern = "%s.liquid")
      @page_template = Liquid::Template.parse("{% include 't_page' with config:config, page:page %}")
      Liquid.cache_classes = false
    elsif @config['mode'] == 'transpile'
      require_relative 'liquid/transpiler'
      Transpiler.new.transpile_dir( @config['liquid'], 'Transpiled', @config['liquid'])
      require( @config['liquid'] + '/Transpiled.rb')
    end
  end

  def dimensions( key)
    @config['dimensions'][key]
  end

  def error( path, msg)
    @errors = @errors + 1
    puts "*** #{msg} [#{path}]"
  end

  def errors?
    @errors > 0
  end

  def get_config( key)
    @config[key]
  end

  def load_words
    if dir = @config['spell_checking']
      unless File.exist?( dir)
        raise "Spell checking directory not found"
      end
      words, names = {}, {}

      Dir.entries( dir).each do |f|
        if /\.txt$/ =~ f
          IO.readlines( dir + '/' + f).each do |line|
            word = line.strip
            next if word == ''

            if word.downcase == word
              words[word.downcase] = true
            else
              names[word] = true
            end
          end
        end
      end

      #p ['load_words', words.size, names.size]
      return words, names
    else
      return nil, nil
    end
  end

  def log( message)
    puts message
  end

  def lookup( path, ref)
    if File.exist?( @source + path) && (! File.directory?( @source + path))
      return path.gsub( /\.txt$/, '.html')
    end
    matches = @key2paths[path]

    if matches.size < 1
      return nil, "Path not found for #{ref}"
    elsif matches.size > 1
      return nil, "Ambiguous path for #{ref}"
    else
      return matches[0], nil
    end
  end

  def offsite?( url)
    if url.index( @config['root_url'])
      false
    elsif /^http/ =~ url
      true
    else
      false
    end
  end

  def prepare( parents, article)
    debug_hook( article, "Preparing")

    begin
      article.prepare( self, parents)
    rescue Exception => bang
      article.error( bang.message)
      raise
    end

    article.children.each do |child|
      prepare( parents + [article], child)
    end
  end

  def record( path)
    @generated[path] = true
  end

  def record_article( article, path)
    if @generated[path]
      article.error( "Duplicate path: #{path} already used")
    else
      @generated[path] = true
    end
  end

  def regenerate( parents, article)
    debug_hook( article, "Regenerating")

    to_data = article.to_data( self, parents)
    params  = {'config' => @config,
               'page'   => to_data}
    if @config['mode'] == 'transpile'
      html = Transpiled.test( params)
    else
      html = @page_template.render( params)
    end

    if /Liquid error:/m =~ html
      article.error( 'Liquid templating error')
    end

    html = article.post_process_html( @config['root_url'], html)
    path = sink_filename( article.filename)
    record_article( article, path)
    Dir.mkdir( File.dirname( path)) if not File.exist?( File.dirname( path))
    rewrite = ! File.exist?( path)

    if ! rewrite
      current = IO.readlines( path).collect {|line| line.chomp}
      rewrite = (current.join("\n").strip != html.strip)
    end

    if rewrite
      puts "... Writing #{path}"
      File.open( path, 'w') {|io| io.print html}
    end

    article.children.each do |child|
      regenerate( parents + [article], child) if child.is_a?( Article)
    end
  end

  def remember( key, path)
    key = key.split('/')
    (0...key.length).each do |i|
      @key2paths[ key[i..-1].join('/')] << path
    end
  end

  def report_errors( article)
    article.report_errors( self)
    article.children.each do |child|
      report_errors( child) if child.is_a?( Article)
    end
  end

  def report_new_words
    if @new_words.size > 0
      @new_words = @new_words.uniq.sort
      puts "... #{@new_words.size} unknown words"
      File.open( @config['misspelt_words'], 'w') do |io|
        io.puts @new_words.join("\n")
      end
    end
  end

  def sink_filename( file)
    if m = /^(\/resources\/)(.*)$/.match( file)
      raise 'resource reference: ' + file
    end
    @sink + ((/^\// =~ file) ? '' : '/') + file
  end

  def source_filename( file)
    @source + ((/^\// =~ file) ? '' : '/') + file
  end

  def spell_check( article, text)

    # Check for spelling checking active
    return unless @words

    # Convert HTML entity codes for letters
    text = text.gsub( /&(.)(acute|caron|cedil|grave|circ|slash|uml|ring);/) do |element|
      element[1..1]
    end

    # Scan scrunched text
    pos, scanner = -100, StringScanner.new( text)
    while pos < scanner.pos
      pos  = scanner.pos
      if scanner.scan_until( /(?:^|^'|\W'|["\s\(])([a-z][a-z'\-]*[a-z])(?:$|\W|\))/i)
        word = scanner[1]

        if m = /^(.*)('s|-like|-wise|-ish|-style)$/.match( word)
          word = m[1]
        end

        if m = /^(pre-|quasi-|ex-|half-|mini-|multi-|non-|pseudo-)(.*)$/i.match( word)
          word = m[2]
        end

        next if @names[word]
        next if @words[word.downcase]

        @new_words << word
        article.error( "Unknown word: #{scanner[1]}")
      end
    end
  end

  def tidy_up( path)
    keep = false
    Dir.entries( path).each do |f|
      next if /^\./ =~ f
      path1 = path + '/' + f
      if File.directory?( path1)
        if tidy_up( path1)
          keep = true
        else
          puts "... Deleting #{path1}"
          File.delete( path1 + '/.DS_Store') if File.exist?( path1 + '/.DS_Store')
          Dir.rmdir( path1)
        end
      else
        if @generated[path1]
          keep = true
        else
          puts "... Deleting #{path1}"
          File.delete( path1)
        end
      end
    end
    keep
  end

  def to_class( name)
    # if name.downcase == 'stdprn'
    #   puts 'DEBUG100'
    # end
    require_relative( 'elements/' + name.downcase)
    if name.downcase == name
      name = name.split( '_').collect {|n| n.capitalize}.join('')
    end
    Kernel.const_get( 'Elements::' + name)
  end
end

compiler = Compiler.new( * ARGV)
compiler.compile
exit 1 if compiler.errors?
